app/com/lynxanalytics/biggraph/frontend_operations/ImportOperations.scala

```diff
@@ -151,27 +145,85 @@ class ImportOperations(env: SparkFreeEnvironment) extends OperationRegistry {
     }
   })
 
-  register("Import Neo4j")(new ImportOperation(_) {
+  register("Import from Neo4j", List(), List("graph"))(new SimpleOperation(_) with Importer {
+    import play.api.libs.json
+    import com.lynxanalytics.biggraph.controllers._
+
     params ++= List(
-      Param("node_label", "Node label"),
-      Param("relationship_type", "Relationship type"),
-      NonNegInt("num_partitions", "Number of partitions", default = 0),
-      Choice("infer", "Infer types", options = FEOption.noyes),
-      Code("sql", "SQL", language = "sql"),
-      Param("imported_columns", "Columns to import"),
-      Param("limit", "Limit"),
-      ImportedTableParam("imported_table"),
+      Param("url", "Neo4j connection", defaultValue = "bolt://localhost:7687"),
+      Param("username", "Neo4j username", defaultValue = "neo4j"),
+      Param("password", "Neo4j password", defaultValue = "neo4j"),
+      Param("vertex_query", "Vertex query", defaultValue = "MATCH (node) RETURN node"),
+      Param("edge_query", "Edge query", defaultValue = "MATCH ()-[rel]->() RETURN rel"),
+      ImportedDataParam(),
       new DummyParam("last_settings", ""))
 
-    def getRawDataFrame(context: spark.sql.SQLContext) = {
-      Neo4jUtil.read(
-        context,
-        params("node_label"),
-        params("relationship_type"),
-        splitParam("imported_columns").toSet,
-        params("infer") == "yes",
-        params("limit"),
-        params("num_partitions").toInt)
+    override def getOutputs(): Map[BoxOutput, BoxOutputState] = {
+      params.validate()
+      assert(params("imported_table").nonEmpty, "You have to import the data first.")
+      assert(!areSettingsStale, areSettingsStaleReplyMessage)
+      import CommonProjectState._
+      val state = json.Json.parse(params("imported_table")).as[CommonProjectState]
+      Map(context.box.output(context.meta.outputs(0)) -> BoxOutputState.from(state))
+    }
+
+    override def runImport(env: com.lynxanalytics.biggraph.BigGraphEnvironment): ImportResult = {
+      import org.apache.spark.sql.functions._
+      import org.apache.spark.sql.types
+      import com.lynxanalytics.biggraph.graph_util.Scripting._
+      val spark = SQLController.defaultContext()(env.sparkDomain)
+      val project = new RootProjectEditor(CommonProjectState.emptyState)
+      def runQuery(query: String, column: String) = {
+        import spark.implicits._
+        val json = spark.read.format("org.neo4j.spark.DataSource")
+          .option("authentication.type", "basic")
+          .option("authentication.basic.username", params("username"))
+          .option("authentication.basic.password", params("password"))
+          .option("url", params("url"))
+          .option("schema.strategy", "string")
+          .option("query", query)
+          .load
+          .persist(org.apache.spark.storage.StorageLevel.MEMORY_AND_DISK)
```

Comment:
I'm still not convinced we need this. Now (if we can switch to using `VerticesToEdges` below) we do seem to get away with a single scan through the results. Yes, spark need to do something more to figure out the schema, but I don't think we should be trying to support that on our end by caching, it should really be spark's responsibility to do something smart about that. And even if it doesn't do anything smart, this is too late to help him here: by the time this persist is executed the schema is figured out, no?

So at the end of the day I believe that now we add all the data to memory/disk when we read through it and then we never read it again.

Reply:
> I'm still not convinced we need this. Now (if we can switch to using `VerticesToEdges` below) we do seem to get away with a single scan through the results. Yes, spark need to do something more to figure out the schema, but I don't think we should be trying to support that on our end by caching, it should really be spark's responsibility to do something smart about that.

Spark doesn't take responsibility for that.

```
scala> val r = Seq("{\"x\": 1}").toDF.rdd.map(x => { println("quack"); x.getString(0) })                                                                               
r: org.apache.spark.rdd.RDD[String] = MapPartitionsRDD[31] at map at <console>:23

scala> spark.read.json(r).show
quack
quack
+---+
|  x|
+---+
|  1|
+---+
```

One quack for figuring out the schema, another quack for getting the data.

> And even if it doesn't do anything smart, this is too late to help him here: by the time this persist is executed the schema is figured out, no?

No. This doesn't do anything yet. `spark.read.json()` (two lines below) figures out the schema. Then the LynxKite operations do the read at some point down the line.

It's not even just a performance question. The data may change in Neo4j between our two reads. (I guess this is an argument against the two separate queries too, but let's not worry about that now.)